File: binary_search/k_closest_element_in_sorted_array.java
// 658. Find K Closest Elements
// https://leetcode.com/problems/find-k-closest-elements/description/

public class Solution {
    public List<Integer> findClosestElements(int[] nums, int k, int target) {
        List<T> arr = Arrays.asList(nums);

        int n = arr.size();
        if (target <= arr.get(0)) {
            return arr.subList(0, k);
        } else if (arr.get(n - 1) <= target) {
            return arr.subList(n - k, n);
        } else {
            int index = Collections.binarySearch(arr, target);
            if (index < 0) // index = -1
                index = -index - 1; // index = -(-1) -1 = 1-1 = 0
            int low = Math.max(0, index - k - 1), high = Math.min(arr.size() - 1, index + k - 1);

            while (high - low > k - 1) {
                if (low < 0 || (target - arr.get(low)) <= (arr.get(high) - target))
                    high--;
                else if (high > arr.size() - 1 || (target - arr.get(low)) > (arr.get(high) - target))
                    low++;
                else
                    System.out.println("unhandled case: " + low + " " + high);
            }
            return arr.subList(low, high + 1);
        }
    }
}

File: binary_search/k_closest_element_in_sorted_array.py
from typing import List
from bisect import bisect_left

class Solution:
    def findClosestElements(self, arr: List[int], k: int, x: int) -> List[int]:
        n = len(arr)

        if x < arr[0]:
            return arr[:k]
        elif x > arr[-1]:
            return arr[-k:]
        else:
            l, r = 0, len(arr) - k

            while l < r:
                mid = (l + r) // 2

                if abs(arr[mid + k] - x) < abs(arr[mid] - x):
                    l = mid + 1
                else:
                    r = mid

        return arr[l : l + k]


# 658. Find K Closest Elements
# https://leetcode.com/problems/find-k-closest-elements/description/


# [1,2,3,4,5,9,10]
#  l
#  r

# k=4, target=3
# l = 0, r = 3
# mid = 1


# l 0, r = 1
# mid = 0
